### PR TITLE
ENH add noncvxpro solver

### PR DESCRIPTION
**Pull request by tomMoral**
This PR goal is to add the solver proposed in [Smooth Bilevel Programming for Sparse Regularization](https://arxiv.org/pdf/2106.01429.pdf) by Poon and Peyre.

The idea is to first go for a naive implementation and then improve stuff.
Code associated with the original paper can be found here: https://github.com/gpeyre/2021-NonCvxPro

To do:

- [ ] verify that the naive implementation converges.
- [ ] add an efficient implementation for fat design matrices.

### CODE REVIEW

File: solvers/noncvx_pro.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    def run(self, n_iter):
+        X, y, lmbd = self.X, self.y, self.lmbd
+        n_features = X.shape[1]
+
+        def objfn(u):
+            Xu = X * u
+            M = lmbd * np.eye(n_features) + np.dot(Xu.T, Xu)
+
+            v1 = np.linalg.solve(M, Xu.T @ y)
```
Comment (agramfort) [inline solvers/noncvx_pro.py:26]:
```suggestion
            if n_samples <= n_features:
                M = lmbd * np.eye(n_features) + np.dot(Xu.T, Xu)
                v1 = np.linalg.solve(M, Xu.T @ y)
            else:
                M = lmbd * np.eye(n_samples) + np.dot(Xu, Xu.T)
                v1 = Xu.T @ np.linalg.solve(M, @ y)
```

File: solvers/noncvx_pro.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        lbfgs_res = sciop.minimize(
+            objfn, u0, method='L-BFGS-B', jac=True, options=myopts
+        )
+        u1 = lbfgs_res.x
+        Xu1 = X * u1
+        M = lmbd * np.eye(n_features) + np.dot(Xu1.T, Xu1)
+
+        v1 = np.linalg.solve(M, Xu1.T @ y)
+
```
Comment (agramfort) [inline solvers/noncvx_pro.py:48]: same here

**Comment by mathurinm** [comment]
I'd like to have a look at it next week

**Comment by mathurinm** [comment]
Just found an implementation by Gabriel himself here: https://github.com/gpeyre/numerical-tours/blob/master/python/optim_7_noncvx_pro.ipynb

**Comment by mathurinm** [comment]
I implemented what is, to me, the best way to solve (A.T @ A + lmbda Id) x = b  without forming A.T @ A (which has cubic cost)
It's still a WIP, and the results are not great. In particular, I don't know if there are numerical errors or what
image


When I plug a `X, y, _ = make_correlated_data(500, 500)` on Gabriel's notebook I get a harsh figure:
image

**Comment by mathurinm** [comment]
I did a few tweaks, maybe a bit better: 
image

maybe @pierreablin can weigh in

**Comment by mathurinm** [comment]
Found an implementation here: https://github.com/gpeyre/2021-NonCvxPro/blob/main/code/solvers.py#L78

**Comment by mathurinm** [comment]
I have tried Gabriel's implementation under the name "noncvx-pro2", that would be enough IMO.
The issue is that there are some runs for which the solver takes an unreasonable time:
image

I cannot reproduce it outside of benchopt (see `test_noncvx.py` that I pushed): across repetitions, the time on the same data is coherent:
image


To see it, run 
`BENCHOPT_DEBUG=1 benchopt run . -s celer -f noncvx-pro  -d simulated -r10 --timeout 1000` and see that for some runs of  "noncvx-pro2", it's fast, while for other it seems to get stuck.

**Comment by mathurinm** [comment]
@tomMoral @agramfort "noncvx-pro2" (green)  follows the implementation by Gabriel here: 
https://github.com/gpeyre/numerical-tours/blob/master/python/optim_7_noncvx_pro.ipynb

It fixes the convergence issue that the initial version had..
The version "noncvx-pro2" seems better, hence if you're ok we can delete the other one, rename and merge.

Results for n_samples = 500, n_features = 200:
||||
|---|---|---|
|image|image|image|

n_samples = 200, n_features = 500
||||
|---|---|---|
|image|image|image|

I chose not to support sparse matrices because this PR has been hanging for a long time now. I think it can be added later.

**Review by agramfort — APPROVED** [review]

File: datasets/simulated.py [hunk 3]
```diff
@@ -12,10 +12,10 @@ class Dataset(BaseDataset):
     # the cross product for each key in the dictionary.
     parameters = {
         'n_samples, n_features': [
-            (100, 5_000),
-            (100, 10_000)
+            (200, 500),
+            (500, 200)
         ],
-        'rho': [0, 0.6],
+        'rho': [0.6],
```
Comment (tomMoral) [inline datasets/simulated.py:18]: not sure why change the parameters? to make default run quicker?

File: objective.py [hunk 4]
```diff
@@ -8,8 +8,9 @@ class Objective(BaseObjective):
     name = "Lasso Regression"
 
     parameters = {
-        'fit_intercept': [True, False],
         'reg': [.5, .1, .05],
+        # 'reg': [.5],
+        'fit_intercept': [False]
```
Comment (tomMoral) [inline objective.py:13]: Same, maybe debug?

File: solvers/noncvx_pro2.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    stopping_strategy = 'iteration'
+
+    def set_objective(self, X, y, lmbd, fit_intercept):
+        self.X, self.y, self.lmbd = X, y, lmbd
+
+    def skip(self, X, y, lmbd, fit_intercept):
+        if fit_intercept:
+            return True, f"{self.name} does not handle fit_intercept"
+        if issparse(X):
+            return True, f"{self.name} does not support sparse design matrices"
```
Comment (tomMoral) [inline solvers/noncvx_pro2.py:23]: Agreed, we can add it later.
```suggestion
        # XXX: make this solver work with sparse matrices.
        if issparse(X):
            return True, f"{self.name} does not support sparse design matrices"
```

File: solvers/noncvx_pro2.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        if n_samples < n_features:
+            def u_opt(v):
+                S = X @ np.diag(v**2) @ X.T + lmbd * np.eye(n_samples)
+                return v * (X.T @ np.linalg.solve(S, y))
+
+            def nabla_f(v):
+                u = u_opt(v)
+                f = 1/(2 * lmbd) * norm(X @ (u*v) - y)**2 + \
+                    (norm(u)**2 + norm(v)**2) / 2
+                g = u * (X.T @ (X @ (u*v) - y)) / lmbd + v
```
Comment (tomMoral) [inline solvers/noncvx_pro2.py:41]: You can make this more efficient by factorizing out the computation of the residual no?
```suggestion
                res = X @ (u*v) - y
                f = 1/(2 * lmbd) * norm(res)**2 + \
                    (norm(u)**2 + norm(v)**2) / 2
                g = u * (X.T @ res) / lmbd + v
```

File: solvers/noncvx_pro2.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            def u_opt(v):
+                T = np.outer(v, v) * C + lmbd * np.eye(n_features)
+                return np.linalg.solve(T, v * Xty)
+
+            def nabla_f(v):
+                u = u_opt(v)
+                x = u * v
+                E = (C@x) @ x + y2 - 2 * x @ Xty
+                f = 1/(2*lmbd) * E + (norm(u)**2 + norm(v)**2)/2
+                g = u * (C @ x - Xty) / lmbd + v
```
Comment (tomMoral) [inline solvers/noncvx_pro2.py:57]: Not sure if it is worth it but:
```suggestion
                Cx = C @ x
                E = Cx @ x + y2 - 2 * x @ Xty
                f = 1/(2*lmbd) * E + (norm(u)**2 + norm(v)**2)/2
                g = u * (Cx - Xty) / lmbd + v
```

**Review by tomMoral — COMMENTED** [review]
A few comments but overall, this LGTM

File: datasets/simulated.py (hunk 3 above)
Comment (mathurinm) [reply datasets/simulated.py:18]: yes but it does not belong to this PR, I'll revert

File: objective.py (hunk 4 above)
Comment (mathurinm) [reply objective.py:13]: yes my bad

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thx @tomMoral

**Comment by tomMoral** [comment]
and thx @mathurinm !

**Pull request merged**